osu.Game/Screens/OnlinePlay/OnlinePlaySongSelect.cs

```diff
@@ -115,45 +114,47 @@ protected override void LoadComplete()
+            {
+                // Enable all free mods if we're creating a new playlist item.
+                // Todo: This needs to be scheduled because mods aren't available until the nested LoadComplete(). Can we do this any better?
+                SchedulerAfterChildren.Add(() => FreeMods.Value = freeModSelect.AllAvailableMods.Where(state => state.ValidForSelection.Value).Select(state => state.Mod).ToArray());
+            }
+
             freeModSelectOverlayRegistration = OverlayManager?.RegisterBlockingOverlay(freeModSelect);
         }
 
         private void onFreestyleChanged(ValueChangedEvent<bool> enabled)
         {
-            if (enabled.NewValue)
-            {
-                freeModsFooterButton.Enabled.Value = false;
-                freeModsFooterButton.Enabled.Value = false;
-                ModsFooterButton.Enabled.Value = false;
+            // If all free mods were previously selected, we'll need to reselect what may now be a larger selection.
+            bool allFreeModsSelected = FreeMods.Value.Count > 0 && freeModSelect.AllAvailableMods.Count(state => state.ValidForSelection.Value) == FreeMods.Value.Count;
 
-                ModSelect.Hide();
-                freeModSelect.Hide();
+            // Remove invalid mods and display the newly available mod panels.
+            Mods.Value = Mods.Value.Where(isValidGlobalMod).ToArray();
+            ModSelect.IsValidMod = isValidGlobalMod;
+            FreeMods.Value = FreeMods.Value.Where(isValidFreeMod).ToArray();
+            freeModSelect.IsValidMod = isValidFreeMod;
 
-                Mods.Value = [];
-                FreeMods.Value = [];
-            }
-            else
-            {
-                freeModsFooterButton.Enabled.Value = true;
-                ModsFooterButton.Enabled.Value = true;
-            }
+            // Reselect all free mods if they were all previously selected (prefer keeping free mods enabled).
+            if (allFreeModsSelected)
+                FreeMods.Value = freeModSelect.AllAvailableMods.Where(state => state.ValidForSelection.Value).Select(state => state.Mod).ToArray();
         }
 
-        private void onModsChanged(ValueChangedEvent<IReadOnlyList<Mod>> mods)
+        private void onGlobalModsChanged(ValueChangedEvent<IReadOnlyList<Mod>> mods)
         {
-            FreeMods.Value = FreeMods.Value.Where(isValidFreeMod).ToList();
-
-            // Reset the validity delegate to update the overlay's display.
+            // Remove incompatible free mods and display the newly available mod panels.
+            FreeMods.Value = FreeMods.Value.Where(isValidFreeMod).ToArray();
             freeModSelect.IsValidMod = isValidFreeMod;
         }
 
         private void onRulesetChanged(ValueChangedEvent<RulesetInfo> ruleset)
         {
-            FreeMods.Value = Array.Empty<Mod>();
+            // Todo: We can probably attempt to preserve across rulesets like the global mods do.
+            FreeMods.Value = [];
```

Comment:
How about this one?

Reply:
The code here isn't changed, only the todo is added and intended more as a future note than anything else. In a previous revision I'd removed this code because its function is somewhat hidden.